### DISCUSSION TITLE
Unexplained Re-rendering

### DISCUSSION THREAD

**Discussion by ProgrammerSM** [react-hook-form/react-hook-form#3723] [Q&A]
I just added react-hook-forms to a gatsby project I'm working on and the behavior I'm experiencing is not what I think it is meant to have. The form seems to be getting re-rendered when I fill out a few of the fields and induce an error to appear. Once this happens it is not maintaining the valid field values but is instead completely wiping out the form. In addition to this behavior, if at that point I try and fill out one of the errored fields it is re-rendered once again and is wiping out the form. I added a render counter to the page, and it seems to re-render a lot more than I find in the various demos on your site or in other playground sites.

I then added the 'watch' prop to the page and a console.log for the page. Now, it re-renders no matter what you enter in any field.

**To Reproduce**

1. Go to 'https://naughty-clarke-b7beca.netlify.app/contact'
2. open your dev tools
3. try to fill in a field

As I can't seem to reproduce this behavior in a playground, I'm not sure how to provide a place to identify the problem.

**Expected behavior**
I would like the form to keep the state of valid fields onSubmit and onBlur and the form to not re-render as often as it is.

**Screenshots**
![image](https://user-images.githubusercontent.com/17866082/102580334-9f439300-40bb-11eb-83c6-fea0285bef5e.png)

**Desktop (please complete the following information):**

- OS: Windows
- Browser Chrome
- Version 87.0.4280.88

**Additional context**

**Current Code**
`import React from 'react'
import { useForm } from 'react-hook-form'

let renderCount = 0;

...
renderCount++
...
const {
  register,
  watch,
  handleSubmit,
  errors
} = useForm();

const allFields = watch()
console.log(allFields)

const onSubmit = data => {
  console.log(data)
}
...
<p>Render Count: {renderCount}</p>
<FormWrapper>
  <form
    onSubmit={handleSubmit(onSubmit)}
  > 

    <input type='hidden' name='form-name' value={formName} ref={register} />
    
    <div className='bot-field'>
      <label htmlFor='bot-field'>Don’t fill this out if you’re human</label>
      <input
        name='bot-field'
        ref={register}
      />
    </div>
    
    <div className='input-field'>
      <label htmlFor='firstName'>First Name</label>
      <input
        name='firstName'
        ref={register({
          required: true,
          maxLength: 80,
        })}
      />
      {errors.firstName && <p className='error'>First name is required</p>}
    </div>

    <div className='input-field'>
      <label htmlFor='lastName'>Last Name</label>
      <input
        name='lastName'
        ref={register({
          required: true,
          maxLength: 100
        })}
      />
      {errors.lastName && <p className='error'>Last name is required</p>}
    </div>

    <div className='input-field'>
      <label htmlFor='email'>Email</label>
      <input
        name='email'
        type='email'
        ref={register({
          required: true,
          pattern: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
        })}
      />
      {errors.email && <p className='error'>Email is required</p>}
    </div>

    <input type='submit'/>
  </form>`

**Comment by bluebill1049** [comment]
your components are getting unmounted while typing, please provide a CSB.

**Reply by bluebill1049** [reply]
did you forget to move the component into its own? eg

```
cont Test = () => {
  cosnt Component = () => {}
  
  render <Component />
}
```

**Reply by ProgrammerSM** [reply]
Ah, got it, I did do that, but your response caused me to think of something that I had not considered. I am using emotion.io for styling the component, and that is what was causing the re-rendering. Thanks for your help
